### PR TITLE
perf(gemma4): walk gathered prompts through shared chunk steps

### LINKED ISSUE #928: gemma4: a whole prompt holds every live stream behind one step
## Description

Admission runs an accepted prompt as one mixed step, however long it is. Every live stream's next token waits out that whole step — at ~0.2 ms per prompt row, a 4K prompt is most of a second of inter-token stall — and at high concurrency each newcomer's first token queues behind the whole prompts ahead of it. The step cost model is a ~27 ms floor plus the per-row term, so the floor argues for gathering short bursts, while the per-row term argues for never letting one step grow long: the fix is to bound the step, not the gather.

## Proposed Plan

- Add an opt-in knob that walks admitted prompts through shared segment steps of at most N prompt rows each; live streams then advance one token per segment instead of waiting out whole prompts.
- Keep the gather: up to the usual prompt cap, newcomers walk the same segment rounds, each round filling one chunk budget across walkers.
- Admit each prompt's pages whole up front; each round's attention plan covers exactly the walked frontier.
- A mid-walk segment's sampled row is discarded; only a prompt's final segment emits its first token and can stop it.

## Acceptance Criteria

- Token-exact equivalence: a gathered chunked walk produces the same sequences as the plain serial path, with and without live decode rows present.
- The KV ledger after a segmented admission matches the whole-prompt admission (frontier, released-front origin, resident page count).
- The knob refuses malformed values at startup; unset keeps whole-prompt steps.

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Closes #928

Admission used to run an accepted prompt as one mixed step, however long. Every live stream's next token waited out that whole step — at ~0.2 ms per prompt row, a 4K prompt is most of a second of inter-token stall — and a newcomer's first token queued behind the whole prompts ahead of it.

`PEGAINFER_MIX_CHUNK_TOKENS=N` (off | 64 ≤ N < 8192, malformed values refuse startup) walks admitted prompts through shared segment steps of at most N prompt rows each. The gather still collects newcomers up to the prompt cap; each round fills one chunk budget across the walkers, the riders' decode rows join every round, and a walker whose final segment lands in a round emits its first token there — a mid-walk segment's sampled row is discarded and can never stop the request. Once the roster drains mid-walk, the remaining tails feed whole. Pages are admitted whole up front and each round's attention plan is truncated to the walked frontier, with coverage asserts where the plan rows shrink.

The logits-head flow (suppression, sampling, logprobs, emission) is shared between the whole-prompt path and the walk rounds rather than duplicated.

The knob stays off by default: chunk granularity is a long-context trade (segments near 2048 rows pay off at 4K–8K prompts; at ~1K prompts segmentation only adds step floors).

## Test Env

Single GPU (sm_89, x86_64), CUDA 12.9, Gemma 4 12B checkpoint.

## Verification

- Two new gates, both green on this tip: a KV-ledger gate holding a segmented admission to the whole-prompt ledger (same frontier, same released-front origin, same resident page count), and an engine gate holding a gathered chunked walk token-exact against the plain serial path — once with a live decode stream riding the rounds and once from an idle roster.
- Existing gates green on this tip: fmt, builds, clippy `-D warnings` with the CI crate set and `--all-targets`, gemma4 lib tests, ragged determinism, the k=2 mixed oracle, the window-crossing oracle.
- The env parse is exercised both ways: unit tests for accept/refuse, and a live startup with a malformed value refusing to serve (`PEGAINFER_MIX_CHUNK_TOKENS="banana" not recognized`, exit 1).
- Serving A/B, one binary, arms interleaved in reversed pairs, three runs per arm, interval readings.

Sixteen coincident ~3900-token prompts landing on a live decode stream — the target regime:

| metric | whole-prompt (unset) | `N=2048` |
|---|---|---|
| stream flood-phase p99 gap | 855–975 ms | **497–526 ms** (−42…−46%, disjoint) |
| flood TTFT p50 | 7.2–8.3 s | 7.3–9.5 s |
| flood TTFT p99 | 14.3–16.4 s | **12.9–14.5 s** |
| flood wall | 13.6–15.6 s | **12.9–13.8 s** |
| quiet-phase p50 gap | 27.5–27.7 ms | 27.5–27.7 ms |

The same flood at ~1900-token prompts shows the documented granularity trade — a 2048-row round can span two prompts, so a round grows past one whole prompt — which is why the knob defaults off and the guidance pins it to long-context workloads:

| metric | whole-prompt (unset) | `N=2048` |
|---|---|---|
| stream flood-phase p99 gap | **432–468 ms** | 519–537 ms |
| flood TTFT p50 | **3.7–4.0 s** | 4.3–5.0 s |
| flood TTFT p99 | 7.3–8.0 s | **6.9–7.6 s** |

Long-context single-stream and batch decode are unaffected:

| metric | whole-prompt (unset) | `N=2048` |
|---|---|---|
| single-stream TPOT median (~1760-token prompt) | 28.7–28.9 ms | 28.8–28.9 ms |
| c8 throughput | 123.4–126.1 tok/s | 127.3–128.9 tok/s |

### CODE REVIEW

File: pegainfer-gemma4/src/engine.rs [hunk 1]
```diff
@@ -422,6 +422,30 @@ fn validate_request(request: &GenerateRequest, prefix_hit_tokens: usize) -> Resu
 /// sampler-row capacity so a burst still leaves decode rows headroom.
 const MIX_MAX_PROMPTS: usize = 4;
 
+/// The chunked-walk knob: `PEGAINFER_MIX_CHUNK_TOKENS=N` walks admitted
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs:425]: DCO passed.

File: pegainfer-gemma4/src/engine.rs [hunk 2]
```diff
@@ -955,6 +1097,288 @@ impl EngineState {
...
+
+            // This round's slate, in walker order: rows to take per walker
+            // and whether that take completes its prompt.
+            let mut budget = chunk;
+            let mut takes: Vec<Option<(usize, bool)>> = vec![None; walkers.len()];
+            for (wi, w) in walkers.iter().enumerate() {
+                if w.failed {
+                    continue;
+                }
+                let rest = w.request.prompt_tokens.len() - w.offset;
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs:1247]: Fixed. The walk loop now sweeps its walkers between rounds: one whose channel closed is marked failed, stops consuming chunk budget, and its pages return with the walker. The gate grew a third phase — a walker cancelled before the first round is dropped while a rider decodes through the rounds, and the surviving walker's sequence still matches the serial path.

File: pegainfer-gemma4/src/engine.rs [hunk 3]
```diff
@@ -955,6 +1097,288 @@ impl EngineState {
...
+            };
+            match flow {
+                Ok((picked, mut lps, _)) => {
+                    let mut si = 0usize;
+                    for (w, t) in walkers.iter_mut().zip(&takes) {
+                        if let Some((take, last)) = *t {
+                            w.offset += take;
+                            if last {
+                                w.first = Some((picked[si], lps[si].take()));
+                            }
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs:1335]: Fixed. A walker whose final segment lands in a round now graduates at that round's boundary: its first token emits and it joins the decode batch — the same capture, stop and emission contract as the whole-prompt form, hoisted out of the walk's former join tail — so an early finisher decodes through the remaining rounds instead of idling, and the walk ends with nothing left to join. The walk gate and the serving A/B were re-run on this shape.

File: pegainfer-gemma4/src/engine.rs [hunk 4]
```diff
@@ -422,6 +422,30 @@ fn validate_request(request: &GenerateRequest, prefix_hit_tokens: usize) -> Resu
 /// sampler-row capacity so a burst still leaves decode rows headroom.
 const MIX_MAX_PROMPTS: usize = 4;
 
+/// The chunked-walk knob: `PEGAINFER_MIX_CHUNK_TOKENS=N` walks admitted
+/// prompts through shared segment steps of at most `N` prompt rows each —
+/// live streams then advance one token per segment instead of waiting out
+/// a whole prompt. Unset (or `off`/`0`) keeps whole-prompt steps.
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs:428]: Done

**Comment by FeathBow** [comment]
@codex review again plz

File: pegainfer-gemma4/src/engine.rs [hunk 5]
```diff
@@ -717,7 +867,10 @@ impl EngineState {
                         None => self.serve.alloc_kv(),
                     };
                     let new_tokens = cand_len - cand_kv.local.seq_len();
-                    if rows_budget + new_tokens > MIX_GATHER_ROWS {
+                    // Chunked mode walks the gather in shared segments that
+                    // pace themselves, so only the count, slot, pool and
+                    // attempt bounds apply.
+                    if self.mix_chunk.is_none() && rows_budget + new_tokens > MIX_GATHER_ROWS {
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs]: Fixed. The local pool provisions exactly one full-context transient on top of the other rows' window-capped footprint, and the chunked gather could reserve up to four — held across many rounds while live rows grow toward their windows. The gather now bounds its upfront reservations to that one provisioned transient, chunked or not (the 512-row ceiling stays the tighter bound for the whole-prompt form). A new gate holds the scenario live: ten one-page streams plus four ~5900-token prompts under a 64-row chunk all run to their budgets — the unbounded gather starved the streams' next pages mid-walk.

**Comment by FeathBow** [comment]
@codex review again plz
